File: 3072/main.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Println(resultArray([]int{2, 1, 3, 3}))
	fmt.Println(resultArray([]int{5,14,3,1,2}))
	fmt.Println(resultArray([]int{3,3,3,3}))
	fmt.Println(resultArray2([]int{2, 1, 3, 3}))
	fmt.Println(resultArray2([]int{5,14,3,1,2}))
	fmt.Println(resultArray2([]int{3,3,3,3}))
}

func resultArray(nums []int) []int {
	arr1 := []int{nums[0]}
	arr2 := []int{nums[1]}
	for i := 2; i < len(nums); i++ {
		if greaterCount(arr1, nums[i]) > greaterCount(arr2, nums[i]) {
			arr1 = append(arr1, nums[i])
		}
		if greaterCount(arr1, nums[i]) < greaterCount(arr2, nums[i]) {
			arr2 = append(arr2, nums[i])
		}
		if greaterCount(arr1, nums[i]) == greaterCount(arr2, nums[i]) {
			if len(arr1) == len(arr2) {
				arr1 = append(arr1, nums[i])
			} else if len(arr1) < len(arr2) {
				arr1 = append(arr1, nums[i])
			} else {
				arr2 = append(arr2, nums[i])
			}
		}
	}
	arr1 = append(arr1, arr2...)
	return arr1
}

func greaterCount(arr []int, val int) int {
	m := 0
	for _, i := range arr {
		if i > val {
			m++
		}
	}
	return m
}

type BinaryIndexedTree struct {
	tree []int
}

func NewBinaryIndexecTree(n int) *BinaryIndexedTree {
	return &BinaryIndexedTree{tree: make([]int, n+1)}
}

func (bit *BinaryIndexedTree) Add(i int) {
	for i < len(bit.tree) {
		bit.tree[i]++
		i += i & -i
	}
}

func (bit *BinaryIndexedTree) Get(i int) int {
	sum := 0
	for i > 0 {
		sum += bit.tree[i]
		i -= i & -i
	}
	return sum
}

func resultArray2(nums []int) []int {
	n := len(nums)
	sortedNums := make([]int, n)
	copy(sortedNums, nums)
	sort.Ints(sortedNums)
	index := make(map[int]int)
	for i, num := range sortedNums {
		index[num] = i + 1
	}

	arr1, arr2 := []int{nums[0]}, []int{nums[1]}
	tree1, tree2 := NewBinaryIndexecTree(n), NewBinaryIndexecTree(n)
	tree1.Add(index[nums[0]])
	tree2.Add(index[nums[1]])
	for i := 2; i < n; i++ {
		count1 := len(arr1) - tree1.Get(index[nums[i]])
		count2 := len(arr2) - tree2.Get(index[nums[i]])
		if count1 > count2 || (count1 == count2 && len(arr1) <= len(arr2)) {
			arr1 = append(arr1, nums[i])
			tree1.Add(index[nums[i]])
		} else {
			arr2 = append(arr2, nums[i])
			tree2.Add(index[nums[i]])
		}
	}
	return append(arr1, arr2...)
}

File: 3072/main.rs
use std::collections::HashMap;

struct BinaryIndexedTree {
    tree: Vec<i32>,
}

struct Solution{}

impl BinaryIndexedTree {
    fn new(n: usize) -> Self {
        BinaryIndexedTree {
            tree: vec![0; n+1],
        }
    }

    fn add(&mut self, mut i: usize) {
        while i < self.tree.len() {
            self.tree[i] += 1;
            i += i & (!i+1);
        }
    }

    fn get(&self, mut i: usize) -> i32 {
        let mut sum = 0;
        while i > 0 {
            sum += self.tree[i];
            i -= i & (!i + 1);
        }
        sum
    }
}

impl Solution {
    pub fn result_array(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut sorted_nums = nums.clone();
        sorted_nums.sort();
        let mut index = HashMap::new();
        for i in 0..n {
            index.insert(sorted_nums[i], i + 1);
        }

        let mut arr1 = vec![nums[0]];
        let mut arr2 = vec![nums[1]];
        let mut tree1 = BinaryIndexedTree::new(n);
        let mut tree2 = BinaryIndexedTree::new(n);
        tree1.add(index[&nums[0]]);
        tree2.add(index[&nums[1]]);

        for i in 2..n {
            let count1 = arr1.len() as i32 - tree1.get(index[&nums[i]]);
            let count2 = arr2.len() as i32 - tree2.get(index[&nums[i]]);
            if count1 > count2 || (count1 == count2 && arr1.len() <= arr2.len()) {
                arr1.push(nums[i]);
                tree1.add(index[&nums[i]]);
            } else {
                arr2.push(nums[i]);
                tree2.add(index[&nums[i]]);
            }
        }
        arr1.extend(arr2);
        arr1
    }
}

fn main() {
    println!("{:#?}", Solution::result_array(vec![2,1,3,3]))
}